packages/hash/frontend/src/lib/entities.ts

```diff
@@ -1,127 +1,419 @@
-import { UnknownEntity } from "../graphql/apiTypes.gen";
+  UnknownEntity as ApiEntity,
+  EntityType as ApiEntityType,
+  Link as ApiLink,
+  LinkGroup as ApiLinkGroup,
+  LinkedAggregation as ApiLinkedAggregation,
+} from "../graphql/apiTypes.gen";
 
-/* eslint-disable no-param-reassign */
+export type ApiEntityIdentifier = {
+  accountId: string;
+  entityId: string;
+};
 
 /**
- * @todo: refactor to adhere to no-param-reassign
+ * Create an entityId that's a stringified object containing the fields we will need later.
+ * The re-written entityId is what should be sent to the block with any Entity,
+ * including for sourceEntityId, destinationEntityId on links
  */
+export const rewriteEntityIdentifier = ({
+  accountId,
+  entityId,
+}: ApiEntityIdentifier) => JSON.stringify({ accountId, entityId });
 
 /**
- * MUTATES an object in order to:
- * - delete all non-id and non-type metadata fields from the root,
- * UNLESS the second argument is 'false'
- * - adds the contents of its 'properties' field to the root
- * - delete the 'properties' field
- * @returns void
+ * Converts an entity from its GraphQL API representation to its Block Protocol representation:
+ * 1. Only provide 'entityId' and 'entityTypeId' at the top level
+ * 2. Re-write 'entityId' so that it is a stringified object of the identifiers we need (i.e. to include accountId)
  */
-const destructivelyMoveEntityPropertiesToRoot = (
-  entity: Partial<UnknownEntity>,
-  preserveExtraMetadata: boolean = false,
-) => {
-  if (!preserveExtraMetadata) {
-    for (const key of Object.keys(entity)) {
-      if (
-        key !== "id" &&
-        key !== "accountId" &&
-        key !== "entityId" &&
-        key !== "entityType" &&
-        key !== "linkGroups" &&
-        key !== "linkedEntities" &&
-        key !== "linkedAggregations" &&
-        key !== "properties"
-      ) {
-        delete entity[key as keyof UnknownEntity];
-      }
-    }
-  }
-  Object.assign(entity, entity.properties);
-  delete entity.properties;
+export const convertApiEntityToBpEntity = ({
+  accountId,
+  entityId,
+  entityTypeId,
+  properties,
+}: Pick<
+  ApiEntity,
+  "accountId" | "entityId" | "entityTypeId" | "properties"
+>): BpEntity => {
+  if (entityId.includes("{")) {
+    throw new Error(
+      `entityId has already been re-written as a stringified object: ${entityId}`,
+    );
+  }
+  return {
+    entityId: rewriteEntityIdentifier({ accountId, entityId }),
+    entityTypeId,
+    properties,
+  };
 };
 
 /**
- * Creates a deep clone of an entity with the contents of the 'properties'
- * field moved to the root of the cloned object.
+ * Converts entities from their GraphQL API representation to their Block Protocol representation.
+ * @see convertApiEntityToBpEntity
+ */
+export const convertApiEntitiesToBpEntities = (
+  records: Pick<
+    ApiEntity,
+    "accountId" | "entityId" | "entityTypeId" | "properties"
+  >[],
+): BpEntity[] => records.map(convertApiEntityToBpEntity);
```

Comment:
`.map` provides extra arguments. You shouldn't pass functions directly to it that aren't specifically designed with that in mind. 

It seems safe, but someone could add a second `number` typed argument to `convertApiEntityToBpEntity` which isn't supposed to be an array index, and TS won't notice the issue. Whereas if you're explicit about what arguments to pass through, it will

This applies to the other functions here too 

```suggestion
): BpEntity[] => records.map(record => convertApiEntityToBpEntity(record));
```

Reply:
https://github.com/hashintel/hash/pull/752/commits/e2949ae69542baa0bfddf850b7477ba2f25458d6